crates/allocator/src/bump.rs

```diff
@@ -108,12 +115,24 @@ impl InnerAlloc {
                 )
             }
         } else if #[cfg(target_arch = "wasm32")] {
-            const fn heap_start() -> usize {
-                0
+            fn heap_start() -> usize {
+                extern "C" {
+                    static __heap_base: usize;
+                }
+                // # SAFETY
+                //
+                // The `__heap_base` symbol is defined by the wasm linker and is guaranteed
+                // to point to the start of the heap.
+                let heap_start =  unsafe { &__heap_base as *const usize as usize };
+                // if the symbol isn't found it will resolve to 0
+                // for that to happen the rust compiler or linker need to break or change
+                assert_ne!(heap_start, 0, "Can't find `__heap_base` symbol.");
+                heap_start
             }
 
-            const fn heap_end() -> usize {
-                0
+            fn heap_end() -> usize {
+                // Cannot overflow on this architecture
+                core::arch::wasm32::memory_size(0) * PAGE_SIZE
```

Comment:
I think the `heap_end` is a bad name because otherwise, it should be like:

```rust
                Self::heap_start() + core::arch::wasm32::memory_size(0) * PAGE_SIZE
```

Maybe it is better to rename the function into `upper_limit`.

Reply:
Got it, `core::arch::wasm32::memory_size(0)` returns the number of pages for the whole memory, including stack and heap. Then yeah, it makes sense.